### DISCUSSION TITLE
Boot time takes 1min 4sec until USB pops up, not executing main.py on custom compiled MicroPython firmware

### DISCUSSION THREAD

**Discussion by ukicomputers** [micropython/micropython#13366] [STM32 / Pyboard] [answered]
Hi! I have a very strange problem. I am using STM32F437ZG with active clock of 168 MHz, and HSE of 25 MHz. To boot my board, it takes around 1min+ to shows up on my PC (and even boot.py). I even tried disbaling RTC. Also, code from main.py never executes. I am specifically using this MCU module: [https://www.mikroe.com/mcu-card-for-stm32-stm32f437zg](https://www.mikroe.com/mcu-card-for-stm32-stm32f437zg). Also, when executing code, I normally write some double LED blinking code (62 lines), I save it, and press Execute button in Thonny, and after saving code, 10sec+ HW_LED2 blinks (LED for status, not configured by me) and until that, code is not executed. That same blinking LED blinks until board is not started up (until USB doesn't pop up, and I guess, not start boot.py). And does anyone know how to change boot.py from firmware and what prints out when in MP help() is called? Please, I need some help. <br><br> Here's my clock configuration from mpconfigboard.h:

```c
#define MICROPY_HW_CLK_PLLM         (25)                // PLLM
#define MICROPY_HW_CLK_PLLN         (336)               // PLLN
#define MICROPY_HW_CLK_PLLP         (RCC_PLLP_DIV2)     // PLLP
#define MICROPY_HW_CLK_PLLQ         (7)                 // PLLQ
#define MICROPY_HW_CLK_LAST_FREQ    (1)
```
Here's my config from file stm32f4xx_hal_conf.h:
```c
#define HSE_VALUE (25000000)
#define LSE_VALUE (32768)
#define EXTERNAL_CLOCK_VALUE (12288000)

#define HSE_STARTUP_TIMEOUT (100)
#define LSE_STARTUP_TIMEOUT (5000)
```
Any help I will very appriciate. Thanks in advance.

**Comment by andrewleech** [comment]
You mention using Thonny, in many cases where strange behaviour is seen this can be a part of the cause.
It would be worth running your code with mpremote to see if it behaves differently with that.

Your custom build, is it from current master? Is there a manifest file freezing code into the unit? You could also just `os.rename("boot.py", "boot.py.bu")` to move that out of the way in case it's got something causing the delays.

Is the startup delay repeatable? Same every time, or sometimes shorter / longer? There could be issues with it taking time for the pll to lock onto the external clock / crystals.

**Reply by ukicomputers** [reply]
My custom build is right from the master, I used example board repo that MP provided on GH. Yes, there is manifest.py that freezes one directory with one code, and code is just written with some defs for LED's (nothing special). The startup delay is repeatable. Also, boot.py is default file, it does not have anything with delay, and it's not even executed with startup. As I said, whenever MP loads something MICROPY_HW_LED2 blinks, and that is happening on startup for 1min. One difference, when using **mpremote**, code executes immediately, but I still have to wait for board to start up before accessing it. In any case needed, here is code: <br>
mpconfigboard.h
```c
#define MICROPY_HW_BOARD_NAME       "Fusion for ARM v8" // ma bre, sve dev board su vam iste, isti pinovi
#define MICROPY_HW_MCU_NAME         "STM32F43xx"

#define MICROPY_HW_HAS_SWITCH       (1)
#define MICROPY_HW_HAS_FLASH        (1)
#define MICROPY_HW_ENABLE_RNG       (1)
#define MICROPY_HW_ENABLE_RTC       (1)
#define MICROPY_HW_ENABLE_DAC       (1)
#define MICROPY_HW_ENABLE_USB       (1)
#define MICROPY_PY_PYB              (0)

// Clock konfiguracija
// Trenutna konfiguracija za F437ZG (NECTO USB default, porgresan 16 PLLM: 25)

#define MICROPY_HW_CLK_PLLM         (25)                // PLLM
#define MICROPY_HW_CLK_PLLN         (336)               // PLLN
#define MICROPY_HW_CLK_PLLP         (RCC_PLLP_DIV2)     // PLLP
#define MICROPY_HW_CLK_PLLQ         (7)                 // PLLQ
#define MICROPY_HW_CLK_LAST_FREQ    (1)

// RTC konfiguracija    
#define MICROPY_HW_RTC_USE_LSE      (1)
#define MICROPY_HW_RTC_USE_US       (0)
#define MICROPY_HW_RTC_USE_CALOUT   (0)

/////////////////////////////////////////////////////////////// CONFIG INTERFACE START

// UART konfiguracija
// mikroBUS slot 1
#define MICROPY_HW_UART1_TX         (pin_B6)
#define MICROPY_HW_UART1_RX         (pin_B7)

// mikroBUS slot 2
#define MICROPY_HW_UART2_TX         (pin_D5)
#define MICROPY_HW_UART2_RX         (pin_D6)

// mikroBUS slot 3
#define MICROPY_HW_UART3_TX         (pin_D8)
#define MICROPY_HW_UART3_RX         (pin_D9)

// mikroBUS slot 4
#define MICROPY_HW_UART4_TX         (pin_C6)
#define MICROPY_HW_UART4_RX         (pin_C7)

// mikroBUS slot 5
#define MICROPY_HW_UART5_TX         (pin_C6)
#define MICROPY_HW_UART5_RX         (pin_C7)

// USB UART
#define MICROPY_HW_UART5_TX         (pin_C6)
#define MICROPY_HW_UART5_RX         (pin_C7)

// I2C busevi
// mikroBUS slot 1-3
#define MICROPY_HW_I2C1_SCL         (pin_B8)
#define MICROPY_HW_I2C1_SDA         (pin_B9)

// mikroBUS slot 4-5
#define MICROPY_HW_I2C2_SCL         (pin_F1)
#define MICROPY_HW_I2C2_SDA         (pin_F0)

// SPI bus
// mikroBUS slot 1
#define MICROPY_HW_SPI1_NSS         (pin_A4)
#define MICROPY_HW_SPI1_SCK         (pin_A5)
#define MICROPY_HW_SPI1_MISO        (pin_A6)
#define MICROPY_HW_SPI1_MOSI        (pin_B5)

// mikroBUS slot 2
#define MICROPY_HW_SPI2_NSS         (pin_B2)
#define MICROPY_HW_SPI2_SCK         (pin_A5)
#define MICROPY_HW_SPI2_MISO        (pin_A6)
#define MICROPY_HW_SPI2_MOSI        (pin_B5)

// mikroBUS slot 3
#define MICROPY_HW_SPI3_NSS         (pin_B4)
#define MICROPY_HW_SPI3_SCK         (pin_A5)
#define MICROPY_HW_SPI3_MISO        (pin_A6)
#define MICROPY_HW_SPI3_MOSI        (pin_B5)

// mikroBUS slot 4
#define MICROPY_HW_SPI4_NSS         (pin_C13)
#define MICROPY_HW_SPI4_SCK         (pin_C10)
#define MICROPY_HW_SPI4_MISO        (pin_C11)
#define MICROPY_HW_SPI4_MOSI        (pin_C12)

// mikroBUS slot 5
#define MICROPY_HW_SPI5_NSS         (pin_D7)
#define MICROPY_HW_SPI5_SCK         (pin_C10)
#define MICROPY_HW_SPI5_MISO        (pin_C11)
#define MICROPY_HW_SPI5_MOSI        (pin_C12)

// CAN bus
#define MICROPY_HW_CAN1_TX          (pin_G12)
#define MICROPY_HW_CAN1_RX          (pin_G13)

/////////////////////////////////////////////////////////////// CONFIG INTERFACE END

// USRSW
#define MICROPY_HW_USRSW_PIN        (pin_B15)
#define MICROPY_HW_USRSW_PULL       (GPIO_NOPULL)
#define MICROPY_HW_USRSW_EXTI_MODE  (GPIO_MODE_IT_FALLING)
#define MICROPY_HW_USRSW_PRESSED    (0)

// Indikacione LED, koristi se samo LED2 za indikaciju boot-a
// dodaj LED1, mozda ne moze da pristupi PB8 ili je PD0?
#define MICROPY_HW_LED2             (pin_B8) // lupio, koristi se za prikaz ucitavanja
#define MICROPY_HW_LED_ON(pin)      (mp_hal_pin_high(pin)) // koristi se hal_ll_gpio
#define MICROPY_HW_LED_OFF(pin)     (mp_hal_pin_low(pin))

// USB konfiguracija, MicroPython se ucitava na USB izlazu na dev ploci
#define MICROPY_HW_USB_FS              (1)
#define MICROPY_HW_USB_VBUS_DETECT_PIN (pin_A9)
```

mpconfigboard.mk
```Makefile
# MCU podesavanja
MCU_SERIES = f4

# Izabracemo ime za F405 mcu, svi isto funkcionisu iz F4 serije, provereno
CMSIS_MCU = STM32F439xx
#AF_FILE = $(BOARD_DIR)/linker_autogen.csv
AF_FILE = boards/stm32f439_af.csv
LD_FILES = $(BOARD_DIR)/linker.ld boards/common_ifs.ld

TEXT0_ADDR = 0x08000000
TEXT1_ADDR = 0x08020000

# MicroPython podesavanja (nista specijalno)
MICROPY_FLOAT_IMPL = double
MICROPY_VFS_LFS1 = 1
MICROPY_VFS_LFS2 = 1

# Moduli u MicroPython-u:
# import <ime_fajla_biblioteke>
FROZEN_MANIFEST = $(BOARD_DIR)/manifest.py
```

stm32f4xx_hal_conf.h
```c
#ifndef MICROPY_INCLUDED_STM32F4XX_HAL_CONF_H
#define MICROPY_INCLUDED_STM32F4XX_HAL_CONF_H

// Oscilatorske vrednosti u Hz
#define HSE_VALUE (25000000)
#define LSE_VALUE (32768)
#define EXTERNAL_CLOCK_VALUE (12288000)

// Oscilatorska cekanja u ms
#define HSE_STARTUP_TIMEOUT (100)
#define LSE_STARTUP_TIMEOUT (5000)

#include "boards/stm32f4xx_hal_conf_base.h"

#endif // MICROPY_INCLUDED_STM32F4XX_HAL_CONF_H
```

manifest.py
```python
# dodavanje python biblioteka
freeze("$(BOARD_DIR)/../../builtin")
```

Sorry if comments are written in strange language. And just one note, I tried disabling RTC, RAM check init, disabling safe mode, but nothing of tried works to fasten up boot time. However, thanks for your reply!

**Reply by andrewleech** [reply]
It could be a filesystem issue, maybe try a build with `MICROPY_HW_HAS_FLASH 0` to see if that changes anything?

You say something starts immediately with mpremote but you still have to wait for startup? I don't quite follow here, it sounds like it works better? Or is there still minute delay?

I'd normally only have LFS2 enabled, LFS1 would generally only be used for compatibility with existing systems. It might be worth disabling that in case it's confusing something.

**Reply by ukicomputers** [reply]
Sorry if I didn't write question generally well! Yes, you follow good :), when executing code with mpremote, code starts almost immediatly. I changed some settings, disabled some things (pyb module) in mpconfigboard.h, and I fasten up boot time to 13 seconds (and I don't know how, it's not big change in code, but OK). <br><br> After some testing, I think same that it is a filesystem issue. Sometimes, I cannot save files, and sometimes when saved, it saves whole file, and after that, it will fill file with bunch of \0. Also sometimes, it saves normally file. **One note:** I have LFS1 **and** LFS2 enabled, so that may be the problem. I will try changing what you said, and I will write a reply. Thanks!

**Reply by ukicomputers** [reply]
So, here is conclusion (for your reply):

- When disabling LFS1, same boot time of 13 seconds again happen (with no changes), only .HEX file size dropped about 33 KB
- When disabling `MICROPY_HW_HAS_FLASH`, same boot time of 13 seconds again happend (with no changes)

In fact, here's how `MICROPY_HW_LED1` blinks until all loads (LED used as showcase for loading), if that helps at all:

https://github.com/micropython/micropython/assets/84191191/1b7f533f-5c7a-4cdc-9142-878bc553b93a

**Comment by andrewleech** [comment]
If you can get a swd debugger connected that can be used to directly find where the delay is too!

**Reply by ukicomputers** [reply]
I have CODEGRIP, but it's closed source, so it's not so useful. And I cannot debug in CODEGRIP's suite, only when writing code with  specific compiler, in specific (NECTO) IDE.

**Comment by ukicomputers** [comment, marked as answer]
The problem was USRSW is pulled low, so board will boot in "safe mode", and that needs prepare, and because of that, the boot time will be long, and main.py or even boot.py will not be executed.

**Reply by andrewleech** [reply]
Ah, yes that's a trap for new boards certainly! I completely forgot about that, good find!
